validator/client/src/test/java/tech/pegasys/artemis/validator/client/DutySchedulerTest.java

```diff
@@ -0,0 +1,198 @@
+    verifyNoMoreInteractions(validatorApiChannel);
+  }
+
+  @Test
+  public void shouldRetryWhenRequestingDutiesFails() {
+    final SafeFuture<Optional<List<ValidatorDuties>>> request1 = new SafeFuture<>();
+    final SafeFuture<Optional<List<ValidatorDuties>>> request2 = new SafeFuture<>();
+    when(validatorApiChannel.getDuties(UnsignedLong.ONE, VALIDATOR_KEYS))
+        .thenReturn(request1)
+        .thenReturn(request2);
+
+    validatorClient.onSlot(compute_start_slot_at_epoch(UnsignedLong.ONE));
+    verify(validatorApiChannel, times(1)).getDuties(UnsignedLong.ONE, VALIDATOR_KEYS);
+
+    request1.completeExceptionally(new RuntimeException("Nope"));
+    assertThat(asyncRunner.hasDelayedActions()).isTrue();
+    asyncRunner.executeQueuedActions();
+
+    // Should retry request
+    verify(validatorApiChannel, times(2)).getDuties(UnsignedLong.ONE, VALIDATOR_KEYS);
+
+    // And not have any more retries scheduled
+    assertThat(asyncRunner.hasDelayedActions()).isFalse();
+  }
+
+  @Test
+  public void shouldScheduleBlockProposalDuty() {
+    final UnsignedLong blockProposerSlot = UnsignedLong.valueOf(5);
+    final ValidatorDuties validator1Duties =
+        ValidatorDuties.withDuties(
+            VALIDATOR1_KEY, 5, 3, List.of(blockProposerSlot), UnsignedLong.valueOf(7));
+    when(validatorApiChannel.getDuties(eq(UnsignedLong.ONE), any()))
+        .thenReturn(SafeFuture.completedFuture(Optional.of(List.of(validator1Duties))));
+
+    final BlockProductionDuty blockCreationDuty = mock(BlockProductionDuty.class);
+    when(dutyFactory.createBlockProductionDuty(validator1, blockProposerSlot))
+        .thenReturn(blockCreationDuty);
+
+    // Load duties
+    validatorClient.onSlot(compute_start_slot_at_epoch(UnsignedLong.ONE));
+
+    // Execute
+    validatorClient.onBlockProductionDue(blockProposerSlot);
+    verify(blockCreationDuty).performDuty();
+  }
+
+  @Test
+  public void shouldNotPerformDutiesForSameSlotTwice() {
+    final UnsignedLong blockProposerSlot = UnsignedLong.valueOf(5);
+    final ValidatorDuties validator1Duties =
+        ValidatorDuties.withDuties(
+            VALIDATOR1_KEY, 5, 3, List.of(blockProposerSlot), UnsignedLong.valueOf(7));
+    when(validatorApiChannel.getDuties(eq(UnsignedLong.ONE), any()))
+        .thenReturn(SafeFuture.completedFuture(Optional.of(List.of(validator1Duties))));
+
+    final BlockProductionDuty blockCreationDuty = mock(BlockProductionDuty.class);
+    when(dutyFactory.createBlockProductionDuty(validator1, blockProposerSlot))
+        .thenReturn(blockCreationDuty);
+
+    // Load duties
+    validatorClient.onSlot(compute_start_slot_at_epoch(UnsignedLong.ONE));
+
+    // Execute
+    validatorClient.onBlockProductionDue(blockProposerSlot);
+    verify(blockCreationDuty).performDuty();
+
+    // Somehow we triggered the same slot again.
+    validatorClient.onBlockProductionDue(blockProposerSlot);
```

Comment:
This functionality is not yet implemented although you're still testing, right?

Reply:
Scheduling the duty is implemented, just the `BlockProductionDuty` doesn't actually do anything yet.

Reply:
I mean the anticorruption layer.

Reply:
I'm not quite sure what you mean.  The anticorruption layer is converting beacon chain events to drive the validator timing channel.  It *shouldn't* wind up triggering the same slot twice but I don't want to depend on that.

Reply:
I'm being particularly defensive here because I know what drives the timing events is likely to change and I'm not sure what it will wind up being yet so making as few assumptions as possible.

Reply:
Oh, I was just curious where the mechanism to make sure we don't propose two blocks in the same slot, but I see that we pass in the slot to `performDuty` so it makes sense now.

Reply:
Yeah I broke that test when I introduced performDutyForSlot because I wrote it with `get` instead of `remove`.  So it's proved it's worth. :)